# Customer Segmentation in a Lisbon Hotel Chain

The objective is to explore the historical customer information of a 4Star Hotel in Lisbon in order to segment customers and discover the distinguishing features of each group.  
  
This should allow market to have a better understanding of customers groups in order to better engage with the customer. These informations may impact several areas of interaction with the customer, eg:
Marketing: channels, timings, reinforcement points, selling points,...
Sales: Pricing , customer value,...
Reception: Types of interaction,...



## DataSet Description

Talk about the variables


## Setup and Import

In [1]:
import os
import csv
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import category_encoders as ce
import collections
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA

Data Loading and Initial Analysis

In [6]:
ds = pd.read_csv(r'data\dataset.csv', sep=";")
ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 111733 entries, 0 to 111732
Data columns (total 29 columns):
ID                      111733 non-null int64
Nationality             111733 non-null object
Age                     107561 non-null float64
DaysSinceCreation       111733 non-null int64
NameHash                111733 non-null object
DocIDHash               110732 non-null object
AverageLeadTime         111733 non-null int64
LodgingRevenue          111733 non-null float64
OtherRevenue            111733 non-null float64
BookingsCanceled        111733 non-null int64
BookingsNoShowed        111733 non-null int64
BookingsCheckedIn       111733 non-null int64
PersonsNights           111733 non-null int64
RoomNights              111733 non-null int64
DistributionChannel     111733 non-null object
MarketSegment           111733 non-null object
SRHighFloor             111733 non-null int64
SRLowFloor              111733 non-null int64
SRAccessibleRoom        111733 non-null int64
SRMe

In [9]:
# Display top 10 rows transposed to show all columns
ds.head(10).transpose()

0  \
ID                                                                    1   
Nationality                                                         PRT   
Age                                                                  52   
DaysSinceCreation                                                   440   
NameHash              0x2C371FD6CE12936774A139FD7430C624F1C4D5109CE6...   
DocIDHash             0x434FD3D59469C73AFEA087017FAF8CA2296493AEABDE...   
AverageLeadTime                                                      59   
LodgingRevenue                                                      292   
OtherRevenue                                                       82.3   
BookingsCanceled                                                      1   
BookingsNoShowed                                                      0   
BookingsCheckedIn                                                     2   
PersonsNights                                                         6   
RoomNights                                                            4   
DistributionChannel                                           Corporate   
MarketSegment                                                 Corporate   
SRHighFloor                                                           0   
SRLowFloor                                                            0   
SRAccessibleRoom                                                      0   
SRMediumFloor                                                         0   
SRBathtub                                                             0   
SRShower                                                              0   
SRCrib                                                                0   
SRKingSizeBed                                                         0   
SRTwinBed                                                             0   
SRNearElevator                                                        0   
SRAwayFromElevator                                                    0   
SRNoAlcoholInMiniBar                                                  0   
SRQuietRoom                                                           0   

                                                                      1  \
ID                                                                    2   
Nationality                                                         PRT   
Age                                                                 NaN   
DaysSinceCreation                                                  1385   
NameHash              0x198CDB98BF37B6E23F9548C56A88B00912D65A9AA0D6...   
DocIDHash             0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...   
AverageLeadTime                                                      61   
LodgingRevenue                                                      280   
OtherRevenue                                                         53   
BookingsCanceled                                                      0   
BookingsNoShowed                                                      0   
BookingsCheckedIn                                                     1   
PersonsNights                                                        10   
RoomNights                                                            5   
DistributionChannel                               Travel Agent/Operator   
MarketSegment                                     Travel Agent/Operator   
SRHighFloor                                                           0   
SRLowFloor                                                            0   
SRAccessibleRoom                                                      0   
SRMediumFloor                                                         0   
SRBathtub                                                             0   
SRShower                                                              0   
SRCrib                                                                0   
SRKingSizeBed      

In [12]:
# Summary statistics for numerical variables
summary=ds.describe(exclude=[np.object])   # exclude only objects, they will be treated next
summary=summary.transpose()  # transpose the summary for easier reading
summary.head(len(summary))

count  unique  \
ID                    111733     NaN   
Nationality           111733     199   
Age                   107561     NaN   
DaysSinceCreation     111733     NaN   
NameHash              111733  107584   
DocIDHash             110732  103480   
AverageLeadTime       111733     NaN   
LodgingRevenue        111733     NaN   
OtherRevenue          111733     NaN   
BookingsCanceled      111733     NaN   
BookingsNoShowed      111733     NaN   
BookingsCheckedIn     111733     NaN   
PersonsNights         111733     NaN   
RoomNights            111733     NaN   
DistributionChannel   111733       4   
MarketSegment         111733       7   
SRHighFloor           111733     NaN   
SRLowFloor            111733     NaN   
SRAccessibleRoom      111733     NaN   
SRMediumFloor         111733     NaN   
SRBathtub             111733     NaN   
SRShower              111733     NaN   
SRCrib                111733     NaN   
SRKingSizeBed         111733     NaN   
SRTwinBed             111733     NaN   
SRNearElevator        111733     NaN   
SRAwayFromElevator    111733     NaN   
SRNoAlcoholInMiniBar  111733     NaN   
SRQuietRoom           111733     NaN   

                                                                    top  \
ID                                                                  NaN   
Nationality                                                         FRA   
Age                                                                 NaN   
DaysSinceCreation                                                   NaN   
NameHash              0x15A713CE687991691A18F6CDC56ABE24979C73CF5D51...   
DocIDHash             0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...   
AverageLeadTime                                                     NaN   
LodgingRevenue                                                      NaN   
OtherRevenue                                                        NaN   
BookingsCanceled                                                    NaN   
BookingsNoShowed                                                    NaN   
BookingsCheckedIn                                                   NaN   
PersonsNights                                                       NaN   
RoomNights                                                          NaN   
DistributionChannel                               Travel Agent/Operator   
MarketSegment                                                     Other   
SRHighFloor                                                         NaN   
SRLowFloor                                                          NaN   
SRAccessibleRoom                                                    NaN   
SRMediumFloor                                                       NaN   
SRBathtub                                                           NaN   
SRShower                                                            NaN   
SRCrib                                                              NaN   
SRKingSizeBed                                                       NaN   
SRTwinBed                                                           NaN   
SRNearElevator                                                      NaN   
SRAwayFromElevator                                                  NaN   
SRNoAlcoholInMiniBar                                                NaN   
SRQuietRoom                                                         NaN   

                       freq         mean        std  min    25%    50%    75%  \
ID                      NaN        55867    32254.7    1  27934  55867  83800   
Nationality           16516          NaN        NaN  NaN    NaN    NaN    NaN   
Age                     NaN      45.6392     17.245  -10     33     47     58   
DaysSinceCreation       NaN      595.027    374.657   36    288    522    889   
NameHash                 75          NaN        NaN  NaN    NaN    NaN    NaN   
DocIDHash              3032          NaN        NaN  NaN    NaN    NaN    NaN   
Aver

In [13]:
# Summary statistics for non numerical variables
summary=ds.describe(include=[np.object],percentiles=None)   
summary=summary.transpose()  
summary.head(len(summary))

count  unique  \
Nationality          111733     199   
NameHash             111733  107584   
DocIDHash            110732  103480   
DistributionChannel  111733       4   
MarketSegment        111733       7   

                                                                   top   freq  
Nationality                                                        FRA  16516  
NameHash             0x15A713CE687991691A18F6CDC56ABE24979C73CF5D51...     75  
DocIDHash            0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...   3032  
DistributionChannel                              Travel Agent/Operator  91019  
MarketSegment                                                    Other  63680

199 Nationalities
We seem to have duplicates on Name Hashes and DocIDHashes , so we may have to do something about this